### DISCUSSION TITLE
Combining | Merging | Extending themes

### DISCUSSION THREAD

**Discussion by lxsymington** [rktjmp/lush.nvim#8] [Q&A] [answered]
First off thanks for a great plugin! It's made configuring a personal colour scheme a pleasure.
I should probably mention that I have minimal knowledge of lua (something I plan to change - but haven't yet). I have also put this under Q&A as I'm less than sure that this functionality isn't either already implemented or trivial to implement.

From my limited understanding, it doesn't look as if there is a way to _extend_ a theme? For example I would like to have variants of my colour scheme (namely light and dark, but doesn't need to be) but there are quite a lot of highlight groups that are the same between them, ideally one could extend a shared base with the different groups. e.g.

Base highlight groups

```lua
{
    ErrorMsg {bg = lightWhite, fg = red, gui = "nocombine"},
    WarningMsg { bg = yellow, fg = black.darken(40), gui = "nocombine,bold,italic" },
    Ignore {},
    Error {bg = red, fg = lightWhite, gui = "nocombine,bold,italic"},
    Todo {bg = blue, fg = lightWhite, gui = "nocombine,bold,italic"},
    -- ... More common highlight groups
}
```

Variant highlight groups (dark)

```lua
{
    Normal {bg = black, fg = white, gui = "nocombine"},
    NormalFloat {bg = grey, gui = "nocombine"},
    NormalNC {bg = lightBlack, fg = white, gui = "nocombine"},
    ColorColumn {bg = black.darken(20), gui = "nocombine"},
    Comment {fg = black.lighten(15), gui = "nocombine,italic"},
    CursorLine {bg = black.darken(20), gui = "nocombine"},
    CursorLineNr {bg = black, fg = yellow, gui = "nocombine,bold"},
    -- ... More variant highlight groups
}
```

similarly for light highlights.

Being able to then combine these to produce a full theme would make maintenance and consistency easier.

Something like:

```lua
lush.merge(base, if background == 'dark' then dark else light end)
```

If this is already trivial, please forgive my lack of understanding and it would be greatly appreciated if you could point me in the right direction!

**Comment by rktjmp** [comment]
TLDR: you can't do it in one step right now but I can plan to (try to) make it work.

This has been something on my mind from the start https://github.com/rktjmp/lush.nvim/blob/main/TODO.md#theme-inheritance

It was some time ago, _pretty_ sure I wrote a demo implementation for that but it was on a different OS install so I think the local branch got nuked. I wasn't sure anyone would even use Lush at the time so I didn't pursue it much further.

I actually _**think**_ the main roadblock was how the API should look, rather than any technical limitation.

You would have to import the spec (base) in the child (dark), then pass the child into lush(), but basically I thought it felt awkward to do that at the end of the spec.

```lua
-- options at end
local base = require('lush_theme.burrito_base')

local spec = lush(function()
  return {
    Normal {...}
  }
end, {inherit: base}) -- awkward when the spec is longer
```

I think I toyed with optionally accepting options as the first argument, but I think I didn't like the complexity it added to maintaining the lush() function, which was feeling pretty overloaded.

```lua
-- optional options at start
local base = require('lush_theme.burrito_base')

rush({parent: base}, function() -- sometimes options, sometimes not
  return {
    Normal {...}
  }
end)
```

Thinking about it now I would like something like,

```lua
-- functionally spec'd
local base = require('lush_theme.burrito_base')

local spec = lush(base).extend(function()
  return {
    Normal {...}
  }
end)
```

You can sneak by, maybe you already know, by requiring your base spec then just accessing it in your second spec, which *sort of* lessens the burden (not really).

```lua
local base = require('lush_theme.burrito_base')

local spec = lush(function()
  return {
    Normal { fg = base.Normal.fg }
  }
end)
```

Ideally you could just pass the group in via the group inheritance stuff,[see doc](https://github.com/rktjmp/lush.nvim/blob/b61956d872d8fd23dd89661d1a54e37bc00786e8/doc/lush.txt#L293) and [demo](https://github.com/rktjmp/lush.nvim/blob/b61956d872d8fd23dd89661d1a54e37bc00786e8/spec/parser_spec.moon#L231) but you have discovered a bug  #9 

You mentioning `merge` does make me wonder though...

If you, thinking out loud,

- write a base spec
- require that spec in dark as a lua module (i.e. a lua table)
- write dark
- use whatever lua's merge table function is (might need external deep merge code, not sure) to merge dark onto base.
- return the merged table to be compiled and applied.

That *might* work for free, no guarantee though.

```lua
local base = require('lush_theme.burrito_base')

local spec = lush(function()
  return {
    Normal {...}
  }
end)

return deep_merge_a_onto_b(base, spec)
```

**Reply by rktjmp** [reply]
You may also find this https://github.com/alaric/nortia.nvim interesting as an exercise in lush (ab)use, though he's doing more than just light/dark.

**Reply by lxsymington** [reply]
Wow! Thanks for such a thorough reply! That's really interesting I'll give your last suggestion a go and see where I get.

Nortia does look really interesting, made me think of the live contrast value (next to the Δ) on the colour picker from https://github.com/lifepillar/vim-colortemplate (it's nice to know that the colour combination works for more than just oneself in the current lighting)
![image](https://user-images.githubusercontent.com/15095115/108548024-2cd17800-72e3-11eb-9cda-8a5d3cf6aa78.png)
Nortia's temporal shifting is a really nice touch too.

**Reply by lxsymington** [reply]
Just wanted to say, I finally got around to giving your last suggestion a go and it worked a treat, so thanks!

**Reply by rktjmp** [reply]
Great. I definitely want to make it easier, just a matter of finding time to write it. At least there is a kind of viable work around now.

**Comment by ellisonleao** [comment]
This would be a nice feature for my [current plugin](https://github.com/npxbr/gruvbox.nvim). I am now using [`vim.tbl_extend`](https://github.com/npxbr/gruvbox.nvim/blob/main/lua/gruvbox/init.lua#L793) to create the whole theme group, by checking for some plugin existence and then extend the base theme table. Not sure if its the right approach, but having this type of feature would definitely make the code a little bit better

**Reply by kunzaatko** [reply]
I unfortunately filed an issue about this before reading through the whole discussion, but this is exactly what I was suggesting. I have some files that define colour groups for [different laguages](https://gist.github.com/kunzaatko/65c2317018afdbfc69abac9c7737722b) (and even different plugins if you would like)

**Reply by rktjmp** [reply]
@ellisonleao I see you had then removed the HL group checks

https://github.com/npxbr/gruvbox.nvim/commit/b938b6d1c2c3cefc35efdd879f420e32d1cbd8ff

What was the reasoning for this? Lush or Vim?

I did wonder if you might hit load order issues with that, if the plugin is hot loaded on a file load you might never get to define the highlight group.

**Reply by ellisonleao** [reply]
hey @rktjmp i've removed it for now until i get all the colors correct in first place, i am still working towards getting them done, the i will get back to the conditional group adding part.

**Comment by rktjmp** [comment]
I have pushed a WIP branch that I would appreciate some feedback on. Does it let you do what you want? I think it probably solves the original post here, as well as  #10 and #13.

https://github.com/rktjmp/lush.nvim/compare/main..extend-spec

The functionality should not change (I believe it covers all it needs to) but the external API (extends, with) might change. (The `{extends: {...}}` is unlikely to change.

Later parents take precedence.

```lua
base = lush(function
  return {
    Normal { ... },
    Function { ... },
    Search { ... }
  })
extras = lush(function
  return {
    Normal { ... },
    Comment { ... },
    Function { ... },
    CursorColumn { Normal },
  })
dark = lush.extends(base, extras).with{function
  return {
    Match { Function },
    CursorLine { CursorColumn },
  })
```

Roughly results in:
```
  Match link extras.Function
  CursorLine link Normal
  Function is extras.Function
  Comment is extras.Comment
  Normal is extras.Normal
  Search is base.Search
```

You can see the basics of how to use it in the test file.
https://github.com/rktjmp/lush.nvim/blob/extend-spec/spec/lush_extends_spec.moon

```lua
lush(function
  return {
    B { A }
  }, {extends: {parent})
```
or 

```lua
lush.extends(parent).with(function 
  return {
    B { A }
  })
```
or

```lua
lush.extends(parent).extends(parent2).with(function 
  return {
    B { A }
  })

-- i.e:

x = lush.extends(parent)
x = x.extends(parent2)
return x.with(function 
  return {
    B { A }
  })
```
or
```lua
lush.extends(parent, parent2).with(function 
  return {
    B { A }
  })
```

or
```lua
parents = {p1, p2}
-- unpack list
lush.extends(unpack(parents)).with(function 
  return {
    B { A }
  })
```

The `extends` chaining isn't hygenic, mostly because lua isn't, so it might change, though I think returning the extends table from files may be useful. Not 100% decided yet, maybe the usage would be basically the same if you just returned specs like normal. Also there is probably too many ways to use it which just makes the API wider to learn.

**Reply by kunzaatko** [reply]
This is great IMO. As for the API getting wider, I think that it isn't necessary so much to have `extend` or maybe even `extends` if there is a `with`.

My main reasoning for this is the desired use-cases:
1) Wanting to define some optional or customizable highlighting in a just a few lines. You don't have to have a settings section and a highlighting section as in elisons `gruvbox.nvim` or even in my theme (I will send it to you shortly). (This is nice because, if you want to have a lots of settings you can have them in a section of a couple tens of LOC and not separated by another 400 lines):

```lua
config['italic_comments'] = vim.g.scheme_italic_comments == 1 and "italic," or ""

theme = lush(function()
    return {
       Normal { ... },
       ... -- lots of LOC
       Comment { ..., gui = config.italic_comments },
       ...
    }end)
```

instead, you can have it defined together... (maybe also in the same closure, so you do not even need to read the variable if you don't need it eventually in the scheme based on some condition (plugin not used or different language))

2) extending costlessly with language specifics:

`scheme.lua` (main):
```lua
local scheme = lush(function()
    return {
        ...
    }
end)

-- this currently does not work (right?), but if we could implement my proposition in #13 (one of the suggested solutions), it could
if vim.api.nvim_get_option('ft') == vim then
   scheme = scheme.with(require'scheme_vim'.scheme_closure)
end
...
return scheme
```

`scheme_vim.lua`
```lua
local scheme_closure = function()
    return {
        ...
    }
end

return {scheme_closure = scheme_closure}
```

I do not see a use-case for extend, where it could not be solved with base lua and `.with(...)`. Is there?

Also there should probably be a convenience function 
```lua
main.with({child1,child2,child3})
```

As for the code, I will look over it as soon as possible (probably in a few days), ok?

**Reply by rktjmp** [reply]
> As for the API getting wider, I think that it isn't necessary so much to have extend or maybe even extends if there is a with.
> I do not see a use-case for extend, where it could not be solved with base lua and .with(...). Is there?

It is entirely necessary to have `extends` because without it, `with` has no function. They're part of the same API.  :eyes: :eyes:

I would like to narrow this thread down onto theme extension only, more specifically around the initial, and probably pretty common use case @lxsymington outlined around having a light and dark theme, which I believe they could now do (with the branch). I think @ellisonleao might want to follow a similar path with gruvbox?

There is a demo of how *I* would approach it here: https://github.com/rktjmp/lush-light-dark-example, some things to note:

- I think it's probably most logical to break your colours out into a separate module that returns a table of HSL values, then access those in all your files. Alternatively you could define them as groups in the base spec (` Red500 { fg: hsl(...)}`), then access them through that (`base.Red500.fg`) though this will pollute vims highlights (not *really* an issue) and is a bit semantically confused.

- You can pass any group of a required spec into your new groups. So if you just want to have a *group* inherit from another spec's _group_ without totally extending the other spec you can.

```lua
local gruv = require 'lush_theme.gruvbox' 

-- no extends(...) call
return lush(function()
  return {
    Normal { ... },
   -- really like gruv comment but want it italic, and I _only_ want the comment style
   Comment { gruv.Comment, gui = "italic" },
  -- alternatively, lighten the bg a bit
  -- Comment { fg = gruv.Comment.fg, bg = gruv.Comment.bg.li(10) }
  }
end)

**Reply by kunzaatko** [reply]
> It is entirely necessary to have extends because without it, with has no function. They're part of the same API. eyes eyes

OK, I think I see it now. I did not understand the API correctly.

**Reply by ellisonleao** [reply]
>You can pass any group of a required spec into your new groups. So if you just want to have a group inherit from another spec's group without totally extending the other spec you can.

That's kinda what i've been wanting. I am thinking about having multiple lua groups files for each language/plugin, then will combine then in the main one using the inheritance support. That gruvbox example is definitely something i would use when its available

**Reply by rktjmp** [reply]
Something like this?

```lua
-- undo_tree.lua
local base = require 'lush_theme.gruv_base' 

return lush(function()
  return {
   UndoTreeHeading { fg = base.Normal.fg, bg = base.Normal.bg },
   -- note: it *should* be ok to write links here, but whether they work is 
   --       dependent on you actually including the base group in the
   --       final spec, else it ends up linking to nothing.
   UndoTreeTime { base.Comment }
  }
end)
```

```lua
-- gruvbox.lua
local base = require 'lush_theme.gruv_base'

specs = {base}
for _, plugin in ipairs(config.plugins) do
  -- where plugins is something like {"undo_tree", "elixir", "fzf"}
  table.insert(specs, require('lush_theme.plugins.'..plugin))
end

return lush.extends(specs).with((function() return {} end)
```

**Reply by kunzaatko** [reply]
The template files are now ready and I think that they adhere to this API (that is not yet in mainline). I added you both (@rktjmp , @ellisonleao ) to the private repo, with the templates.

**Reply by ellisonleao** [reply]
thanks @kunzaatko i will be live streaming on wednesday and i hope to make those changes over there (twitch.tv/npxbr btw)

**Comment by rktjmp** [comment, marked as answer]
`extends.with` and  `merge` now in mainline cc5ba4038e6b300a97dadf2b0f38c6d06505ccf7.  https://github.com/rktjmp/lush.nvim#advanced-usage.
